src/sentry/integrations/pagerduty/integration.py

```diff
@@ -0,0 +1,180 @@
+            project_ids_and_services = data.pop("project_mapping")
+
+            with transaction.atomic():
+                PagerDutyServiceProject.objects.filter(pagerduty_service__in=self.services).delete()
+
+                for p_id, s in project_ids_and_services.items():
+                    # create the record in the table
+                    project = Project.objects.get(pk=p_id)
+                    service = PagerDutyService.objects.get(id=s["service"])
+                    PagerDutyServiceProject.objects.create(
+                        project=project, pagerduty_service=service
+                    )
+
+    def get_config_data(self):
+        config = self.org_integration.config
+        project_mappings = PagerDutyServiceProject.objects.filter(
+            pagerduty_service__in=self.services
+        )
+        data = {}
+        for pm in project_mappings:
+            data[pm.project_id] = {"service": pm.pagerduty_service_id}
+        config = {}
+        config["project_mapping"] = data
+        return config
+
+    @property
+    def services(self):
+        services = PagerDutyService.objects.filter(organization_integration=self.org_integration)
+
+        return services
+
+
+class PagerDutyIntegrationProvider(IntegrationProvider):
+    key = "pagerduty"
+    name = "PagerDuty"
+    metadata = metadata
+    features = frozenset([IntegrationFeatures.ALERT_RULE])
+    integration_cls = PagerDutyIntegration
+
+    setup_dialog_config = {"width": 600, "height": 900}
+
+    def get_pipeline_views(self):
+        return [PagerDutyInstallationRedirect()]
+
+    def post_install(self, integration, organization):
+        services = integration.metadata["services"]
+        try:
+            org_integration = OrganizationIntegration.objects.get(
+                integration=integration, organization=organization
+            )
+        except OrganizationIntegration.DoesNotExist:
+            return
+
+        with transaction.atomic():
+            for service in services:
+                PagerDutyService.objects.create_or_update(
+                    organization_integration=org_integration,
+                    integration_key=service["integration_key"],
+                    service_id=service["id"],
+                    service_name=service["name"],
+                )
+
+    def build_integration(self, state):
+        config = json.loads(state.get("config"))
+        account = config["account"]
+        # PagerDuty gives us integration keys for various things, some of which
+        # are not services. For now we only care about services.
+        services = filter(lambda x: x["type"] == "service", config["integration_keys"])
+
+        return {
+            "name": account["name"],
+            "external_id": account["subdomain"],
+            "metadata": {"services": services},
+        }
+
+
+class PagerDutyInstallationRedirect(PipelineView):
+    def get_app_url(self):
+        app_id = options.get("pagerduty.app-id")
```

Comment:
Should this throw an exception if the `app_id` isn't set? I bet an on-premise user will bump into this.

Reply:
I set the default to be `""` - is that not enough?

Reply:
Yeah the code will work as it is, but on-premise folks are prone to misconfigure sentry and needing help to solve those issues.